Fix release builds

Release builds have been broken for a long time, I should fix it one day…

It's caused by the fact that we use `#[cfg(debug_assertions)]` for a parameter in the `invite_member` route, and Rocket's `#[post(…)]` macro doesn't expect it. I was hoping it would be found and fixed by the Rocket team, but seems like they never noticed it.

One day I should take some time to create a minimal reproducible example for them or just remove `#[cfg(debug_assertions)]` as a temporary workaround.

```log
   Compiling prose-pod-api v0.2.0 (/project)
error[E0061]: this function takes 7 arguments but 8 arguments were supplied
  --> src/v1/invitations/routes.rs:60:21
   |
60 | pub(super) async fn invite_member<'r>(
   |                     ^^^^^^^^^^^^^
...
68 |     #[cfg(debug_assertions)] invitation_service: LazyGuard<UnauthenticatedInvitationService<'r>>,
   |                              ------------------ unexpected argument of type `LazyGuard<unauthenticated_invitation_service::UnauthenticatedInvitationService<'_>>`
   |
note: function defined here
  --> src/v1/invitations/routes.rs:60:21
   |
60 | pub(super) async fn invite_member<'r>(
   |                     ^^^^^^^^^^^^^
61 |     conn: Connection<'r, Db>,
   |     ------------------------
62 |     app_config: &State<AppConfig>,
   |     -----------------------------
63 |     server_config: LazyGuard<ServerConfig>,
   |     --------------------------------------
64 |     jid: LazyGuard<BareJid>,
   |     -----------------------
65 |     notifier: LazyGuard<Notifier<'r>>,
   |     ---------------------------------
66 |     invitation_controller: LazyGuard<InvitationController<'r>>,
   |     ----------------------------------------------------------
67 |     req: Json<InviteMemberRequest>,
   |     ------------------------------

For more information about this error, try `rustc --explain E0061`.
error: could not compile `prose-pod-api` (lib) due to 1 previous error
```

--- a/.github/workflows/release.yaml
+++ b/.github/workflows/release.yaml
@@ -97,7 +97,3 @@ jobs:
           platforms: |
             linux/amd64
             linux/arm64
-          # FIXME: Fix release builds (https://github.com/prose-im/prose-pod-api/issues/54)
-          #   and then remove the `--debug` flag.
-          build-args: |
-            CARGO_INSTALL_EXTRA_ARGS=--debug

--- a/.github/workflows/test.yaml
+++ b/.github/workflows/test.yaml
@@ -74,10 +74,13 @@ jobs:
             target
           key: ${{ runner.os }}-cargo
 
+      - name: Try building in release mode
+        run: cargo check --release
+
       - name: Install SQLite3
         run: sudo apt-get update && sudo apt-get install -y sqlite3
 
-      - name: Test code
+      - name: Test code in debug mode
         run: cargo test
 
   integration-test:
@@ -128,8 +131,10 @@ jobs:
       - name: Install Task
         uses: arduino/setup-task@v2
 
-      - name: Build images
-        run: scripts/run-locally/build-images
+      - name: Build image
+        # NOTE: We need to build in debug mode because we need to use
+        #   `debug_only` configuration during integration tests.
+        run: task build-image -- --debug
 
       - name: Install SQLite3
         run: sudo apt-get update && sudo apt-get install -y sqlite3

--- a/crates/rest-api/src/features/invitations/invite_member.rs
+++ b/crates/rest-api/src/features/invitations/invite_member.rs
@@ -3,6 +3,7 @@
 // Copyright: 2023–2024, Rémi Bardon <remi@remibardon.name>
 // License: Mozilla Public License v2.0 (MPL v2.0)
 
+#[cfg(debug_assertions)]
 use std::ops::Deref as _;
 
 use rocket::{post, response::status, serde::json::Json, State};
@@ -18,18 +19,39 @@ use service::{
     AppConfig,
 };
 
+#[cfg(not(debug_assertions))]
+use crate::responders::Created;
 use crate::{
     error::prelude::*,
     guards::{Db, LazyGuard},
-    responders::Either,
 };
 #[cfg(debug_assertions)]
 use crate::{
+    features::invitations::guards::UnauthenticatedInvitationService,
     features::members::{rocket_uri_macro_get_member_route, Member},
     forms::JID as JIDUriParam,
+    responders::Either,
 };
 
-use super::{guards::*, model::*, rocket_uri_macro_get_invitation_route};
+use super::{model::*, rocket_uri_macro_get_invitation_route};
+
+#[cfg(not(debug_assertions))]
+pub type InviteMemberResponse = Created<WorkspaceInvitation>;
+#[cfg(not(debug_assertions))]
+fn ok(invitation: WorkspaceInvitation, resource_uri: String) -> InviteMemberResponse {
+    Ok(status::Created::new(resource_uri).body(invitation.into()))
+}
+#[cfg(debug_assertions)]
+pub type InviteMemberResponse = Result<
+    Either<status::Created<Json<WorkspaceInvitation>>, status::Created<Json<Member>>>,
+    Error,
+>;
+#[cfg(debug_assertions)]
+fn ok(invitation: WorkspaceInvitation, resource_uri: String) -> InviteMemberResponse {
+    Ok(Either::left(
+        status::Created::new(resource_uri).body(invitation.into()),
+    ))
+}
 
 #[derive(Serialize, Deserialize)]
 pub struct InviteMemberRequest {
@@ -41,6 +63,7 @@ pub struct InviteMemberRequest {
 }
 
 /// Invite a new member.
+#[cfg(not(debug_assertions))]
 #[post("/v1/invitations", format = "json", data = "<req>")]
 pub async fn invite_member_route<'r>(
     conn: Connection<'r, Db>,
@@ -50,9 +73,56 @@ pub async fn invite_member_route<'r>(
     notifier: LazyGuard<Notifier>,
     invitation_controller: LazyGuard<InvitationController>,
     req: Json<InviteMemberRequest>,
+) -> InviteMemberResponse {
+    invite_member_route_(
+        conn,
+        app_config,
+        server_config,
+        user_info,
+        notifier,
+        invitation_controller,
+        req,
+    )
+    .await
+}
+
+/// Invite a new member and auto-accept the invitation if enabled.
+#[cfg(debug_assertions)]
+#[post("/v1/invitations", format = "json", data = "<req>")]
+pub async fn invite_member_route<'r>(
+    conn: Connection<'r, Db>,
+    app_config: &State<AppConfig>,
+    server_config: LazyGuard<ServerConfig>,
+    user_info: LazyGuard<UserInfo>,
+    notifier: LazyGuard<Notifier>,
+    invitation_controller: LazyGuard<InvitationController>,
+    req: Json<InviteMemberRequest>,
+    invitation_service: LazyGuard<UnauthenticatedInvitationService>,
+) -> InviteMemberResponse {
+    invite_member_route_(
+        conn,
+        app_config,
+        server_config,
+        user_info,
+        notifier,
+        invitation_controller,
+        req,
+        invitation_service,
+    )
+    .await
+}
+
+/// Invite a new member.
+pub async fn invite_member_route_<'r>(
+    conn: Connection<'r, Db>,
+    app_config: &State<AppConfig>,
+    server_config: LazyGuard<ServerConfig>,
+    user_info: LazyGuard<UserInfo>,
+    notifier: LazyGuard<Notifier>,
+    invitation_controller: LazyGuard<InvitationController>,
+    req: Json<InviteMemberRequest>,
     #[cfg(debug_assertions)] invitation_service: LazyGuard<UnauthenticatedInvitationService>,
-) -> Result<Either<status::Created<Json<WorkspaceInvitation>>, status::Created<Json<Member>>>, Error>
-{
+) -> InviteMemberResponse {
     let db = conn.into_inner();
     let server_config = server_config.inner?;
     let notifier = notifier.inner?;
@@ -78,21 +148,21 @@ pub async fn invite_member_route<'r>(
         )
         .await?;
 
-    if cfg!(debug_assertions) && app_config.debug_only.automatically_accept_invitations {
-        let jid = invitation.jid;
-        let resource_uri = uri!(get_member_route(jid.clone().into())).to_string();
-        let member = MemberRepository::get(db, &jid).await?.unwrap();
-        let response: Member = member.into();
-        Ok(Either::right(
-            status::Created::new(resource_uri).body(response.into()),
-        ))
-    } else {
-        let resource_uri = uri!(get_invitation_route(invitation.id)).to_string();
-        let response: WorkspaceInvitation = invitation.into();
-        Ok(Either::left(
-            status::Created::new(resource_uri).body(response.into()),
-        ))
+    #[cfg(debug_assertions)]
+    {
+        if app_config.debug_only.automatically_accept_invitations {
+            let jid = invitation.jid;
+            let resource_uri = uri!(get_member_route(jid.clone().into())).to_string();
+            let member = MemberRepository::get(db, &jid).await?.unwrap();
+            let response: Member = member.into();
+            return Ok(Either::right(
+                status::Created::new(resource_uri).body(response.into()),
+            ));
+        }
     }
+
+    let resource_uri = uri!(get_invitation_route(invitation.id)).to_string();
+    ok(invitation.into(), resource_uri)
 }
 
 // ERRORS

--- a/crates/service/src/features/auth/auth_service.rs
+++ b/crates/service/src/features/auth/auth_service.rs
@@ -6,6 +6,7 @@
 use std::{fmt::Debug, ops::Deref, sync::Arc};
 
 use secrecy::SecretString;
+#[cfg(debug_assertions)]
 use serde::{Deserialize, Serialize};
 
 use crate::{models::BareJid, prosody::ProsodyOAuth2Error};
@@ -63,7 +64,6 @@ pub enum AuthError {
     ProsodyOAuth2Err(#[from] ProsodyOAuth2Error),
     #[error("Invalid credentials")]
     InvalidCredentials,
-    #[cfg(debug_assertions)]
     #[error("{0}")]
     Other(String),
 }

--- a/crates/service/src/features/invitations/invitation_controller.rs
+++ b/crates/service/src/features/invitations/invitation_controller.rs
@@ -10,7 +10,9 @@ use chrono::{DateTime, Utc};
 use rand::{distributions::Alphanumeric, thread_rng, Rng};
 use sea_orm::{DatabaseConnection, DbErr, ItemsAndPagesNumber, ModelTrait as _};
 use secrecy::{ExposeSecret as _, SecretString};
-use tracing::{debug, error, warn};
+#[cfg(debug_assertions)]
+use tracing::warn;
+use tracing::{debug, error};
 
 use crate::{
     dependencies,

--- a/crates/service/src/features/network_checks/models/dns.rs
+++ b/crates/service/src/features/network_checks/models/dns.rs
@@ -39,18 +39,21 @@ pub enum DnsRecord {
     },
 }
 
-/// NOTE: Only used in tests.
-#[cfg(debug_assertions)]
 impl DnsRecord {
-    pub fn record_type(&self) -> DnsRecordDiscriminants {
-        DnsRecordDiscriminants::from(self)
-    }
     pub fn hostname(&self) -> &DomainName {
         match self {
             Self::A { hostname, .. } | Self::AAAA { hostname, .. } | Self::SRV { hostname, .. } => {
                 hostname
             }
         }
+    }
+}
+
+/// NOTE: Only used in tests.
+#[cfg(debug_assertions)]
+impl DnsRecord {
+    pub fn record_type(&self) -> DnsRecordDiscriminants {
+        DnsRecordDiscriminants::from(self)
     }
 }
 

